app/src/ui/rebase/rebase-flow.tsx

```diff
@@ -323,55 +112,33 @@ export class RebaseFlow extends React.Component<
     targetBranch: string,
     commits: ReadonlyArray<CommitOneLine>
   ) => {
-    if (this.state.step.kind !== RebaseStep.ChooseBranch) {
-      throw new Error(`Invalid step to start rebase: ${this.state.step.kind}`)
+    if (this.props.step.kind !== RebaseStep.ChooseBranch) {
+      throw new Error(`Invalid step to start rebase: ${this.props.step.kind}`)
     }
 
-    const totalCommitCount = commits.length
+    this.props.dispatcher.initializeRebaseProgress(
+      this.props.repository,
+      commits
+    )
 
-    const startRebaseAction = async () => {
-      const result = await this.props.dispatcher.rebase(
+    const startRebaseAction = () => {
+      return this.props.dispatcher.rebase(
         this.props.repository,
         baseBranch,
-        targetBranch,
-        {
-          rebasedCommitCount: 0,
-          totalCommitCount,
-          progressCallback: this.updateProgress,
-        }
+        targetBranch
       )
-
-      if (result === RebaseResult.CompletedWithoutError) {
-        this.moveToCompletedState()
-      }
     }
 
-    this.setState(() => {
-      const currentCommitSummary =
-        commits.length > 0 ? commits[0].summary : undefined
-
-      const rebasedCommitCount = 1
-      const newProgressValue = rebasedCommitCount / commits.length
-
-      return {
-        step: {
-          kind: RebaseStep.ShowProgress,
-          rebaseAction: startRebaseAction,
-        },
-        progress: {
-          commits,
-          value: formatRebaseValue(newProgressValue),
-          rebasedCommitCount,
-          currentCommitSummary,
-        },
-      }
+    this.props.dispatcher.setRebaseFlowStep(this.props.repository, {
+      kind: RebaseStep.ShowProgress,
+      rebaseAction: startRebaseAction,
     })
   }
 
   private onContinueRebase = async () => {
-    if (this.state.step.kind !== RebaseStep.ShowConflicts) {
+    if (this.props.step.kind !== RebaseStep.ShowConflicts) {
```

Comment:
Does this line imply that we can only continue a rebase if we need to show conflicts? If so, the doc for this function doesn't make that clear.
> Continue with the rebase after the user has resovled all conflicts with
   tracked files in the working directory.

Reply:
These guard checks were present to help ensure the callers were enforcing the right transitions, and were useful while I was building out the flow, but are less valuable now that things are stable.

I starting down the path in this PR to define a pattern for these callbacks, where the first parameter received is the step that is initiating the transition:

```ts
private showRebaseConflictsBanner = (step: ShowConflictsStep) => {
 ...
}
```

This helps to make clear:

 - what state is initiating the callback
 - that data the callback needs is available in the step
 - that invalid states are no longer representable at compile time, avoiding the need for checks like this at runtime

This makes the props inside a component a bit more complex than plain functions:

```tsx
  readonly onContinueRebase: (step: ShowConflictsStep) => void
  readonly onAbortRebase: (step: ShowConflictsStep) => void
  readonly showRebaseConflictsBanner: (step: ShowConflictsStep) => void
```

I've pushed some more commits to this branch to wrap up this work - let me know how this feels.

Reply:
Overall, this feels nice 👍